Add code and yaml for Istio as an addon

ostromart opened the pull request:
This commit adds the following:
 - YAML for istio components to the addons directory (versions with
   and without auth)
 - logic to copy the appropriate YAML file into the addons manager
   dir, depending on ENABLE_ISTIO and ISTIO_AUTH_TYPE env vars
 - changes to addon manager script which allow plugins to exist in
   namespace other than kube-system. This is needed for Istio
   functionality
 - additional initialization for Istio in addon manager.

This has been manually tested - e2e tests are in progress but require additional code to enable istio in tests. See:
 - https://github.com/istio/istio/pull/2841
 - https://github.com/kubernetes/test-infra/pull/6612

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Allow [Istio](https://istio.io/) to be installed as an addon. This simplifies the Istio setup process on k8s and GKE. Istio team is trying to make it easier for end users to get to a basic functional installation.  
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

ostromart commented:
/verify

ostromart commented:
/assign @mikedanese
/cc @MrHohn

MrHohn commented:
/ok-to-test

MrHohn commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -2218,6 +2218,13 @@ EOF
     update-prometheus-to-sd-parameters ${metadata_proxy_yaml}
   fi
+  if [[ "${ENABLE_ISTIO:-}" == "true" ]]; then
+    if [[ "${ISTIO_AUTH_TYPE:-}" == "MUTUAL_TLS" ]]; then
```
Humm, I was under the impression that we also need to pipe these envs into build-kube-env:
https://github.com/kubernetes/kubernetes/blob/9b646503113ae9eb33e57ca231edb868247ba1f5/cluster/common.sh#L472

Does it just work?

ostromart replied:
I manually tested by just temporarily setting the variables directly in the script itself. This will work for GKE which has an internal path for setting env vars but it will be good to implement the non-GKE path as well. I'll take a look at that.

MrHohn replied:
I see, yeah it would be great to have a common path so that we could enable it in CI jobs as well (for GCE cluster).

ostromart replied:
https://github.com/kubernetes/test-infra/pull/6612 is still in progress from @krzyzacy 
Let me see if I can help with that. Seems like it would be generally useful functionality. 
The public API and CLI are being updated right now (ETA ~2 weeks) so that would be the way to enable it in tests when it becomes available.

MrHohn commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -31,6 +31,7 @@ KUBECTL_OPTS=${KUBECTL_OPTS:-}
 ADDON_PATH=${ADDON_PATH:-/etc/kubernetes/addons}
+ISTIO_PATH=${ADDON_PATH}/istio
 
 SYSTEM_NAMESPACE=kube-system
```
Remove this if no longer needed?

MrHohn commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -193,6 +194,10 @@ for obj in $(find /etc/kubernetes/admission-controls \( -name \*.yaml -o -name \
 
+# Create Istio resources, which are in istio-system namespace.
+log INFO "== Applying Istio addon (if present) at $(date -Is) =="
+${KUBECTL} ${KUBECTL_OPTS} apply -f ${ISTIO_PATH} --recursive | grep -v configured
```
Why do we need this extra step? Are Istio addon manifests not under ${ADDON_PATH}?

ostromart replied:
They are there but it's a long manifest and not all of the resources are labeled with CLUSTER_SERVICE_LABEL, only the deployments at the moment so the apply fails for some resources unless they are already created.

MrHohn replied:
Thanks for the clarification, it sounds like we need a way to express the creation order. I wonder if there is a more generic way to achieve this.

Note that the resources without any of `CLUSTER_SERVICE_LABEL` or `ADDON_MANAGER_LABEL` will not be deleted when its manifest is removed from the folder (e.g. when disabling istio). Addon-manager's cleanup logic is accomplished via `kubectl apply -f {FOLDER} --prune=true -l {LABEL}`, which relies on label matching.

Also, I vaguely remember that current implementation of `kubectl apply --prune` doesn't handle the multi-namespaces cases very well. Let's say we have two manifests, one is under namespace1, the other under namespace2. `kubectl apply -f --prune` works well for cleaning up resource as long as there are at least one resource in each namespace defined in manifests. Once the last resource in that namespace gets remove from the manifest (e.g. again, disabling istio), this resource will not be cleaned up because `kubectl` will not even look into that namespace if it is neither given via `--namespace` flag nor defined in the manifests.

I start feeling that the solution would be messy given the inflexibility of addon-manager...

mikedanese commented on cluster/addons/istio/auth/istio-auth.yaml:
```diff
@@ -0,0 +1,1950 @@
+kind: CustomResourceDefinition
+apiVersion: apiextensions.k8s.io/v1beta1
+metadata:
+  name: apikeies.config.istio.io
```
apikeys?

ostromart replied:
It does seem to be a typo in Istio repo. Still investigating on Istio side but here we have to use the misspelled version since that's what is in Istio repo at the version we are using here.

ostromart replied:
https://github.com/istio/istio/pull/3101

ostromart commented:
/cc @yusuoh

ostromart commented:
/retest

ostromart commented:
Per conversation with @MrHohn I've labeled all Istio resources with addonmanager.kubernetes.io/mode: Reconcile, and the Deployments as kubernetes.io/cluster-service: "true" so these show up in kubectl cluster-info. I've removed the special handling for unlabeled resources out of kube-addons.sh so the only change there is now removing the namespace filter.

ostromart commented:
/retest

MrHohn commented on cluster/addons/istio/noauth/istio.yaml:
```diff
@@ -0,0 +1,2198 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: istio-system
```
As we discussed before, for enable this seems fine. But for disable, addon-manager will not do the cleanup when all resources under this new namespace are removed from manifests. Is it acceptable for the alpha phase?

MrHohn commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -124,12 +124,12 @@ function reconcile_addons() {
 
   log INFO "== Reconciling with addon-manager label =="
-  ${KUBECTL} ${KUBECTL_OPTS} apply --namespace=${SYSTEM_NAMESPACE} -f ${ADDON_PATH} \
+  ${KUBECTL} ${KUBECTL_OPTS} apply -f ${ADDON_PATH} \
```
I think this will work, but I need to confirm how `--prune` behave when multiple namespaces appear in manifest. Will give an update soon.

MrHohn replied:
Seems like it should work with the multiple namespaces case, kubectl records the visited namespaces for pruning:
https://github.com/kubernetes/kubernetes/blob/48388fec7eaad4ac8d84fbe20673ffacf41964a1/pkg/kubectl/cmd/apply.go#L391-L402

Also a bad news, at this moment the object types for pruning is hardcoded so don't think it will work well with the CRDs added by istio...
https://github.com/kubernetes/kubernetes/blob/48388fec7eaad4ac8d84fbe20673ffacf41964a1/pkg/kubectl/cmd/apply.go#L418-L440

MrHohn commented on cluster/addons/istio/noauth/istio.yaml:
```diff
@@ -0,0 +1,2198 @@
+  labels:
+    k8s-app: istio
+    addonmanager.kubernetes.io/mode: Reconcile
+    kubernetes.io/cluster-service: "true"
```
Oops sorry, I said it wrong, this label should be attached to the `istio-mixer` service instead.
https://github.com/kubernetes/kubernetes/blob/c322f1d06504fa5b81e117ec79dd77fb4e769b4f/pkg/kubectl/cmd/clusterinfo.go#L75-L80

ostromart replied:
Done.

MrHohn commented on cluster/addons/istio/noauth/istio.yaml:
```diff
@@ -0,0 +1,2194 @@
+    k8s-app: istio
+    addonmanager.kubernetes.io/mode: Reconcile
+spec:
+  replicas: 1
```
General comment for all deployments that have `replicas: 1`: would any of them want the ability to horizontally scaled up? If so, you may want to remove this field in manifest so addon-manager will not scale it back down.

ostromart replied:
This is ok for alpha since we won't be seeing loads that the default sizing can't handle (in fact it's supposed to be able to handle much higher loads). But longer term we need a solution since we do want to reconcile all the other Deployment fields.

ostromart replied:
Per offline discussion, I've removed the replicas setting so it will not be reconciled.

MrHohn commented on cluster/addons/istio/noauth/istio.yaml:
```diff
@@ -0,0 +1,2194 @@
+    addonmanager.kubernetes.io/mode: Reconcile
+spec:
+  replicas: 1
+  template:
```
Another general comment: it seems like all podTemplates don't have resource quotas set (memory and cpu), so all pods will be in `BestEffort` class. It may increase the risk of overcommitting the nodes and istio components being easily evicted. For important system pod, a general suggestion would be making it `Guaranteed` class and set a high pod priority. Don't think this is a blocker but would be great to have better resource accounting in future.


Ref https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/, https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/

ostromart replied:
I had some discussion in the team meeting about this just now. For alpha, it's ok to go with the BestEffort since we don't yet have a full understanding of performance and scale. This is something that will evolve as we move to 1.0. 
But this is a very good point and we will be tracking this issue.

MrHohn commented on cluster/addons/istio/auth/istio-auth.yaml:
```diff
@@ -0,0 +1,2194 @@
+      - name: grafana-data
+        emptyDir: {}
+---
+---
```
nit: double `---` lines?

ostromart replied:
Done.

MrHohn reviewed (approved):
Looks good, don't have other major comments.
/lgtm

ostromart commented:
/retest

ostromart commented:
@mikedanese, would you please merge this if there are no outstanding concerns?

mikedanese reviewed (commented):
Should these manifests go in cluster/gce/addons if they are gke specific?

Can you squash fixup commits?

ostromart committed:
Add code and yaml for Istio as an addon

This commit adds the following:
 - YAML for istio components to the addons directory (versions with
   and without auth)
 - logic to copy the appropriate YAML file into the addons manager
   dir, depending on ENABLE_ISTIO and ISTIO_AUTH_TYPE env vars
 - changes to addon manager script which allow plugins to exist in
   namespace other than kube-system. This is needed for Istio
   functionality
 - additional initialization for Istio in addon manager.

mikedanese commented:
/lgtm

ostromart commented on cluster/addons/istio/auth/istio-auth.yaml:
```diff
@@ -0,0 +1,2185 @@
+  resources: ["endpoints", "pods", "services"]
+  verbs: ["get", "list", "watch"]
+- apiGroups: [""]
+  resources: ["namespaces", "nodes", "secrets"]
```
1. Squash: done.
2. The manifests themselves are not GKE specific, the part that's GKE specific is installing the istio manifest based on a startup env var which is set by GKE. But a user can also just copy the manifest through some other means and it will work the same. 
3. Sent this question to istio security team, should have answer shortly.

ostromart replied:
I've added you to the email exchange on 3. Is this a blocker for this PR or could we open an issue and move ahead?

ostromart replied:
Just a followup on the secrets issue. Per offline discussion, there are currently only two types of secret that Istio needs to read - however, there's no way to filter based on type so there doesn't seem to be a way to narrow the scope. Unless you have some ideas here I'm not going to pursue this further.

mikedanese commented:
/hold

mikedanese commented:
/unhold
